Rewrite scalar solve to division

jessegrabowski opened the pull request:
<!-- !! Thank your for opening a PR !! -->

<!--- Provide a self-contained summary of your changes in the Title above -->
<!--- This is what will be shown in the automatic release notes: https://github.com/pymc-devs/pytensor/releases -->

## Description
Small optimization that rewrites `solve(a,b) -> b / a` when the core shape of `a` is `(1, 1)`. This avoids calling a LAPACK routine in a case where it's simply not necessary.

This came up in the gradient of `minimize`, for cases where the function being minimized has only one input. The L_op in that case requires a bunch of linear algebra, but it can all be rewritten away when we're just dealing with scalars.

I also tweaked `rewrite_cholesky_diag_to_sqrt_diag` to apply to the (1, 1) case -- this allows `cho_solve` to be rewritten to just b / a ** 2, without any linalg calls. ~~The only bummer is that the `sqr` and `sqrt` don't cancel. I get this graph:~~ 

Here is the graph, including the new sqr/sqrt cancellation:

```
import pytensor.tensor as pt
import pytensor
x = pt.tensor('x', shape=(1, 1))
c_and_lower = pt.linalg.cholesky(x), True
b = pt.tensor('b', shape=(None,))

f = pytensor.function([x, b], pt.linalg.cho_solve(c_and_lower, b))

f.dprint()

Composite{(i1 / abs(i0))} [id A] 2
 ├─ Squeeze{axis=1} [id B] 1
 │  └─ x [id C]
 └─ DimShuffle{order=[x]} [id D] 0
    └─ b [id E]

Inner graphs:

Composite{(i1 / abs(i0))} [id A]
 ← true_div [id F] 'o0'
    ├─ i1 [id G]
    └─ Abs [id H]
       └─ i0 [id I]
```

## Related Issue
<!--- It is good practice to first open an issue explaining the bug / new feature that is addressed by this PR -->
<!--- Please type an `x` in one of the boxes below and provide the issue number after the # sign: -->
- [ ] Closes #
- [ ] Related to #

## Checklist
<!--- Make sure you have completed the following steps before submitting your PR -->
<!--- Feel free to type an `x` in all the boxes below to let us know you have completed the steps: -->
- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)
<!--- You may find this guide helpful: https://mainmatter.com/blog/2021/05/26/keeping-a-clean-git-history/ -->

## Type of change
<!--- Select one of the categories below by typing an `x` in the box -->
- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):
<!--- Additionally, if you are a maintainer or reviewer, please make sure that the appropriate labels are added to this PR -->


<!-- readthedocs-preview pytensor start -->
----
📚 Documentation preview 📚: https://pytensor--1453.org.readthedocs.build/en/1453/

<!-- readthedocs-preview pytensor end -->

jessegrabowski commented:
Oh I didn't even think about the domain issue. That's a good reason we don't do that rewrite.

ricardoV94 commented:
> Oh I didn't even think about the domain issue. That's a good reason we don't do that rewrite.

We handle log(exp) and exp(log), fine, it's just that depending on the order the rewrite has a man switch or not

jessegrabowski commented:
Common complex L

jessegrabowski commented on pytensor/tensor/rewriting/linalg.py:
```diff
@@ -1020,3 +1027,42 @@ def slogdet_specialization(fgraph, node):
+
+    # Special handling for different types of solve
+    match core_op:
+        case SolveTriangular():
```
Definitely. I was just being too infatuated with match.

ricardoV94 replied:
Match doesn't work with the class (instead of instance) as an instance check?

jessegrabowski replied:
No. See the [match PEP](https://peps.python.org/pep-0636/#adding-a-ui-matching-objects)

ricardoV94 replied:
Yeah it doesn't instantiate stuff:

```python
class AB:
    def __init__(self):
        print("AB was initialized")

class AA(AB):
    def __init__(self):
        print("AA was initialized")

class BB(AB):
    def __init__(self):
        print("BB was initialized")

aa = AA()
bb = BB()

print("Starting match")
for x in (aa, bb, 1):
    match x:
	case AA():
            print("match aa")
        case BB():
            print("match bb")
        case AB():
            print("match ab")
        case _:
            print("match nothing")

# AA was initialized
# BB was initialized
# Starting match
# match aa
# match bb
# match nothing
```

Not sure how it's implemented, very surprising syntax imo, but seems fine to use like you did

ricardoV94 replied:
Wonder if we could start using it for `match x.owner:  Apply(op=Elemwise(scalar_op=Add))` kind of stuff instead of `x.owner is not None and isinstance(x.owner.op, Elemwise) and isinstance(x.owner.op, Add))` :thinking:

ricardoV94 replied:
https://github.com/pymc-devs/pytensor/issues/1455

jessegrabowski commented:
We had a test that assumed Cholesky of a 1x1 matrix should not be rewritten to `sqrt(diag(x))`, which is obviously wrong. I'm not sure why we had that, but I added a check for that case to the cholesky diag rewrite, and removed the test.

ricardoV94 commented on pytensor/tensor/rewriting/math.py:
```diff
@@ -400,6 +400,39 @@ def local_exp_log(fgraph, node):
+            new_out = cast(new_out, old_out.dtype)
+        return [new_out]
+
+    # Case for sqr(sqrt(x)) -> x
+    if isinstance(prev_op, ps.Sqr) and isinstance(node_op, ps.Sqrt):
+        new_out = x.owner.inputs[0]
+        old_out = node.outputs[0]
+
+        # Handle potential integer to float cast by sqrt
+        if x.dtype != old_out.dtype:
+            new_out = cast(new_out, old_out.dtype)
+        return [new_out]
```
And we register the nan rewrite only in stabilize/specialize so if you have something like `sqrt(sqr(sqrt(x))` you end up with `sqrt(x)` instead of `sqrt(switch(x <=0, nan, x)`, because the `sqr(sqrt(x))` rewrites can act without rush.

ricardoV94 replied:
You would rewrite `sqr(sqrt(x)) -> switch(x <=0, np.nan, x)`

jessegrabowski replied:
Understood

jessegrabowski commented on pytensor/tensor/rewriting/linalg.py:
```diff
@@ -908,6 +910,11 @@ def rewrite_cholesky_diag_to_sqrt_diag(fgraph, node):
 
     [input] = node.inputs
+
+    # Check if input is a (1, 1) matrix
+    if all(input.type.broadcastable[-2:]):
+        return [pt.sqrt(input)]
```
hey man it removes an entire lapack call

ricardoV94 commented on tests/tensor/rewriting/test_math.py:
```diff
@@ -2031,6 +2031,56 @@ def test_exp_log_nested(self, nested_expression, expected_switches):
+        # sqrt(x) ** 2 -> x
+        data = self.rng.normal(size=(4, 3)).astype(config.floatX)
+        x = pt.tensor("x", shape=(None, None))
+        f = function([x], sqr(sqrt(x)), mode=self.mode)
```
For this simple rewrite I would just call rewrite_graph and use `assert_equal_computations` with the expected graph. It's faster and more explicit.

No need to compile a function and evaluate when the expected graph is obviously correct

ricardoV94 commented on tests/tensor/rewriting/test_math.py:
```diff
@@ -2031,6 +2031,66 @@ def test_exp_log_nested(self, nested_expression, expected_switches):
+        out = sqr(sqrt(x))
+        out = rewrite_graph(out, include=["canonicalize", "specialize", "stabilize"])
+
+        fg = FunctionGraph(inputs=[x], outputs=[out])
+
+        assert not any(
+            isinstance(node.op, Elemwise)
+            and isinstance(node.op.scalar_op, ps.Sqr | ps.Sqrt)
+            for node in fg.toposort()
+        )
+        assert any(
+            isinstance(node.op, Elemwise) and isinstance(node.op.scalar_op, ps.Abs)
+            for node in fg.toposort()
+        )
```
I wouldn't call it implicit :)

ricardoV94 commented on tests/tensor/rewriting/test_math.py:
```diff
@@ -2031,6 +2031,47 @@ def test_exp_log_nested(self, nested_expression, expected_switches):
+            and isinstance(node.op.scalar_op, ps.Sqr | ps.Sqrt)
+            for node in nodes
+        )
+        assert f([1, 2, 3]).dtype in ["float32", "float64"]
```
Nit, if you already bother to compile and eval why not check with the expected output? strict=True in np.testing.asser_array_equal enforces dtypes.

But I wouldn't compile either. This check you had is equivalent to asserting the dtype of the output of the rewrite. And/or you can use the equal_computations approach. Approving already anyway, up to you.
